### PR TITLE
[FLINK-31092][sql-gateway][table-common] Fix Hive ITCase fail with OutOfMemoryError

### PR DESCRIPTION
**Pull request by fsk119**
## What is the purpose of the change

*Fix test OOM in the Hive Module. There are two problems in the test:*
*1. Gateway doesn't wait for all tasks finished before releasing the resources. This is because the `FutureTask#cancel`
only interrupts the running worker but doesn't stop the thread.*
*2. ServiceLoaderUtil keeps loading the results even though some unrelated exception happens.*

## Brief change log

- *Commit f016d7c3 adds a case about a task that contains an infinite loop*
- *Commit c888b066 fixes this problem by stopping the thread explicitly*
- *Commit 63e97914 adds a case that `FactoryUtil` keeps loading results*
- *Commit dc846d8b fixes the problem by checking the exception type*

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/ServiceLoaderUtil.java [hunk 1]
```diff
@@ -44,6 +51,20 @@ static <T> List<LoadResult<T>> load(Class<T> clazz, ClassLoader classLoader) {
...
+                    LOG.debug(
+                            "NoClassDefFoundError when loading a "
+                                    + clazz.getCanonicalName()
+                                    + ". This is expected when trying to load a format dependency but no flink-connector-files is loaded.",
+                            t);
+                    // After logging, we just ignore this failure
+                    continue;
+                }
+
+                if (!clazz.isInstance(t)) {
```
Comment (XComp) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/factories/ServiceLoaderUtil.java:64]: imho, this if condition still doesn't make sense. And the error handling should probably live in the calling method `FactoryUtil.discoverFactories`. `ServiceLoaderUtil.load` seems to be more general. I provided a commit show what I had in mind in [42453004](https://github.com/fsk119/flink/pull/1/commits/42453004898473a6e28b78f7603b30c3b1e24951) in [PR fsk119/flink:#1](https://github.com/fsk119/flink/pull/1/). But we could even go one step further and delete `ServiceLoaderUtil` completely. imho, it doesn't add much value to the codebase as it's only called in `FactoryUtil.discoverFactories`.

But see [PR fsk119/flink:#1](https://github.com/fsk119/flink/pull/1/) more like a suggestion. ...it's your call.

**Comment by fsk119** [comment]
I have verify the sql-gateway module in the jdk11 environment with command

```
mvn clean verify -Pjava11
```

and all tests pass.

**Review by wuchong — APPROVED** [review]
LGTM.

**Comment by fsk119** [comment]
All tests pass in my pipeline: https://dev.azure.com/1059623455/Flink/_build/results?buildId=565&view=results

**Pull request closed without merging**